### PR TITLE
Add File Attachment Capability to new_task Tool

### PR DESCRIPTION
**Pull request by daniel-lxs**
# Add File Attachment Capability to new_task Tool

Closes: #2931 

This PR introduces a new parameter to the new_task tool, enabling the parent agent to attach files by mimicking how users reference files in tasks.

## Context

This enhancement adds a `<files>` parameter to the new_task tool. Files specified in this parameter will be processed using the same logic as the `read_file` function, including:

- Application of the maximum line limit constraint
- Inclusion of line numbers in file contents
- Support for specifying line ranges using the format `path:startLine:endLine`

## Purpose:
This allows tasks to be created with immediately available file context, which:

- Avoids redundant file read requests from child agents
- Reduces token consumption from sequential file operations
- Encourages explicit file inclusion in task definitions
- Improves overall task execution efficiency

## Implementation Details

1. **File Specification Format**:
   - Basic format: `<file>path/to/file.ts</file>`
   - With line range: `<file>path/to/file.ts:10:50</file>` (lines 10-50)

2. **Backend Changes**:
   - Added `AttachedFileSpec` interface in `shared/tools.ts`
   - Enhanced `newTaskTool.ts` to parse file paths and extract line ranges
   - Updated `ClineProvider.ts` to pass file specifications to the Cline constructor
   - Refactored `Cline.ts` to process attached files with line ranges and format them appropriately in the initial prompt

3. **UI Enhancements**:
   - Updated `ChatRow.tsx` to display attached files as badges within the task content block
   - Added support for showing line ranges in the badges (e.g., `@filename:10-50`)
   - Clicking the badge opens the file with a selection of the line range if specified

4. **Documentation**:
   - Updated the tool description in `new-task.ts` to explain the new syntax and provide examples

## Visual Comparison

**Before**: New tasks required the child agent to explicitly request files, leading to redundant operations and token usage.

image

**After**: Files are immediately available to the child agent, with badges showing attached files:
- `@filename` for full files
- `@filename:10-50` for files with line ranges

image
image

## Testing Protocol

1. **Basic Functionality**:
   - Create a new task with a single attached file
   - Create a new task with multiple attached files
   - Verify files appear in the initial context of the child agent

2. **Line Range Testing**:
   - Attach a file with a specific line range (e.g., `src/core/Cline.ts:580:600`)
   - Verify only the specified lines are included in the context
   - Test edge cases (start line > end line, line numbers beyond file length)

3. **UI Verification**:
   - Confirm file badges appear correctly in the task approval message
   - Verify line ranges are displayed properly in the badges
   - Check that badges are displayed side-by-side in a flex container
   - Click on the badges to confirm the file opens correctly

4. **Error Handling**:
   - Test with non-existent files
   - Test with invalid line ranges
   - Verify appropriate error messages are shown

----

> [!IMPORTANT]
> This PR adds file attachment capability with line range specifications to the `new_task` tool, updating backend processing and UI display.
> 
>   - **Behavior**:
>     - Adds `<files>` parameter to `new_task` tool for attaching files with optional line ranges.
>     - Files processed using `read_file` logic, including line limits and line numbers.
>   - **Backend Changes**:
>     - Introduces `AttachedFileSpec` interface in `shared/tools.ts`.
>     - Updates `newTaskTool.ts` to parse and handle file paths and line ranges.
>     - Modifies `ClineProvider.ts` to pass file specifications to `Cline`.
>     - Refactors `Cline.ts` to format attached files for task prompts.
>   - **UI Enhancements**:
>     - Updates `ChatRow.tsx` to display attached files as badges with line ranges.
>     - Clicking badges opens files with specified line selections.
>   - **Documentation**:
>     - Updates `new-task.ts` to include new syntax and examples for file attachments.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooVetGit%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for d5a1d4af904060bdb8e5bce3075e3acdf9740847. You can [customize](https://app.ellipsis.dev/RooVetGit/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by daniel-lxs** [comment]
@cte @mrubens This test is unrelated to the feature but it still failed, it seems like a faulty test.

**Comment by adamhill** [comment]
🚀

**Comment by jezweb** [comment]
this will be a brilliant time saver thank you

**Comment by daniel-lxs** [comment]
Rebased, now it's up to date with the recent Cline.ts refactor

**Comment by hannesrudolph** [comment]
@mrubens @cte can we please review this tomorrow?

File: src/core/task/Task.ts [hunk 1]
```diff
@@ -558,27 +565,155 @@ export class Task extends EventEmitter<ClineEvents> {
...
+							actualEndLine0Based = startLine0Based - 1
+						}
+						actualEndLine1Based = actualEndLine0Based + 1
+					}
+				}
+
+				// Add header line with path and line range (using 1-based line numbers for display)
+				attachedFilesContent += `${relativeFilePath}:${startLine1Based}:${actualEndLine1Based}${truncationNotice}\n`
+
+				// Add the file content in a markdown code block
```
Comment (jr) [inline src/core/task/Task.ts:669]: Is there a reason this uses markdown instead of the faux-xml that a lot of the other tools use? (e.g. https://github.com/RooVetGit/Roo-Code/blob/6201c985f532f3cb70672ed39b1d1678d61d6659/src/core/tools/readFileTool.ts#L224). Do the models handle markdown better?

If there isn't a particular reason not to, it would be nice if it was consistent with what readFileTool does.

**Review by jr — COMMENTED** [review]
I like this idea a lot.

It seems like it overlaps a lot with what readFileTool does. Is it practical to share most of that code, or are there reasons I'm not thinking of why that wouldn't work?

**Comment by daniel-lxs** [comment]
@jr 
I agree with you, I can refactor to avoid duplicating the code and make the content read more consistent.

File: src/core/task/Task.ts (hunk 1 above)
Comment (daniel-lxs) [reply src/core/task/Task.ts]: I now use the same formatting for both file reading operations

**Comment by daniel-lxs** [comment]
@jr 
I refactored `readFileTool.ts` to extract the read file logic and put it on a shared file `src/shared/fileReadUtils.ts`, now this shared file works for both this new feature and the `read_file` tool.

I have also split the tests, now we test the file reading logic separately from the `read_file` tool, while keeping the same coverage from the previous tests.

**Comment by KJ7LNW** [comment]
@samhvw8 please review #3086 with respect to #2886:

It looks like this #3086 simply factors out `processFileForReading`, which may not affect #2886 at all, but I want to avoid any complications because #2886 has become quite mature.

Both PRs touch `read_file` and @samhvw8 (plus other users, including myself) have put a lot of testing into #2886 and it works quite well.

File: src/shared/fileReadUtils.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+	return result
+}
+
+/**
+ * Formats a processed file result to a string representation
+ * @param result The processed file result
+ * @returns Formatted string
+ */
+export function formatProcessedFileResultToString(result: ProcessedFileReadResult): string {
```
Comment (KJ7LNW) [inline src/shared/fileReadUtils.ts:231]: This should probably stay in `readFileTool.ts` and export it if you need it somewhere else. 

 someday functions like this will be in a tool base class to build xml without string concatenation with a proper xml builder library .

File: src/core/prompts/tools/new-task.ts [hunk 3]
```diff
@@ -2,22 +2,32 @@ import { ToolArgs } from "./types"
...
+</files>
 </new_task>
 
 Example:
 <new_task>
 <mode>code</mode>
 <message>Implement a new feature for the application.</message>
+<files>
+<file>src/somefile.ts</file>
+<file>src/anotherfile.ts:10:50</file>
```
Comment (KJ7LNW) [inline src/core/prompts/tools/new-task.ts:29]: Please see #2886 and work with @samhvw8 so that are line range formats do not diverge; multiple range formats would probably confuse the model across tools.

Gemini and Claude have both been shown empirically tested that the following format used in #2886 works very well for the model. this is a real life example:

```xml
<read_file>
<args>
  <file>
    <path>src/core/config/__tests__/ContextProxy.test.ts</path>
    <line_range>130-140</line_range> <!-- Around line 134 -->
    <line_range>160-170</line_range> <!-- Around line 166 -->
  </file>
  <file>
    <path>src/exports/api.ts</path>
    <line_range>60-75</line_range> <!-- Around line 67 -->
  </file>
  <file>
    <path>src/exports/types.ts</path> <!-- To see RooCodeSettings definition -->
  </file>
  <file>
    <path>src/schemas/index.ts</path>
    <line_range>150-160</line_range> <!-- To see historyItemSchema definition -->
  </file>
</args>
</read_file>
```

standardizing XML representation will simplify it tool class refactors in the future

File: src/core/tools/readFileTool.ts [hunk 4]
```diff
@@ -52,69 +47,43 @@ export async function readFileTool(
...
 					return
 				}
-
-				// Convert to 0-based index
-				endLine -= 1
-			}
-
-			const accessAllowed = cline.rooIgnoreController?.validateAccess(relPath)
-
-			if (!accessAllowed) {
```
Comment (KJ7LNW) [inline src/core/tools/readFileTool.ts:100]: where did access allowed go?

if it moved somewhere centralized, then great but I want to make sure we are not creating regressions

File: src/core/tools/readFileTool.ts [hunk 5]
```diff
@@ -52,69 +47,43 @@ export async function readFileTool(
...
 
 			if (isFullRead) {
 				// No snippet for full read
-			} else if (startLine !== undefined && endLine !== undefined) {
-				lineSnippet = t("tools:readFile.linesRange", { start: startLine + 1, end: endLine + 1 })
-			} else if (startLine !== undefined) {
-				lineSnippet = t("tools:readFile.linesFromToEnd", { start: startLine + 1 })
-			} else if (endLine !== undefined) {
-				lineSnippet = t("tools:readFile.linesFromStartTo", { end: endLine + 1 })
+			} else if (requestedStartLine !== undefined && requestedEndLine !== undefined) {
```
Comment (KJ7LNW) [inline src/core/tools/readFileTool.ts:81]: These look like unnecessary variable rename that cause diff review bloat. 

Unless there is a really good reason to rename variables let's not do that if we can avoid it, so we can tell what actually change for the pull request, and what is original .

**Comment by daniel-lxs** [comment]
Hey @KJ7LNW, the refactor to the `read_file` logic was made to unify the logic so it can be used in both places that is needed without duplicating the code.

The `read_file` tool unit tests were separated from the file reading logic, but I made sure to avoid modifying the tests themselves on both cases, if you see a case where a test no longer serves its original purpose please let me know.

File: src/core/tools/readFileTool.ts [hunk 6]
```diff
@@ -136,120 +105,23 @@ export async function readFileTool(
...
-					lines = maxReadFileLine
-				}
-
-				const lineRangeAttr = ` lines="1-${lines}"`
-
-				// Maintain exact format expected by tests
-				contentTag = `<content${lineRangeAttr}>\n${content}</content>\n`
-			}
+			// Use shared utility functions for file processing
+			const result = await processFileForReading(
```
Comment (KJ7LNW) [inline src/core/tools/readFileTool.ts:109]: This is a huge function hoist, and we need to make sure there are no hallucinations: please show the diff to a smart model (o4-high, etc) and ask it to validate that there is exactly no behavior change.

Especially, please validate around output formatting and file line endings, leading and trailing blank lines, etc, which have been validated in detail (and hopefully the tests will still enforce). If you were to look up the pull request related to XML output formatting, this was done meticulously for consistent output and help gemini with understanding things.

**Review by KJ7LNW — CHANGES_REQUESTED** [review]
Overall, I like the implementation and I think this will be very useful.

Most of my comments are about best practices and ways to minimize differences across the refactor.

Thank you for putting this together!

File: src/core/tools/readFileTool.ts (hunk 6 above)
Comment (KJ7LNW) [reply src/core/tools/readFileTool.ts:109]: > Hey @KJ7LNW, the refactor to the read_file logic was made to unify the logic so it can be used in both places that is needed without duplicating the code.
> The read_file tool unit tests were separated from the file reading logic, but I made sure to avoid modifying the tests themselves on both cases, if you see a case where a test no longer serves its original purpose please let me know.

If the test still pass, then hopefully it is fine. I only made this comment because I have seen models hallucinate during large code moves, so I always ask the model in a separate question showing only the diff and have it validate that the difference in behavior is identical.

It probably is identical but, I operate on the side of caution with these kinds of things.

**Comment by daniel-lxs** [comment]
This can potentially break power-steering and it's not something we want at the moment.
